dpcpp/matrix/csr_kernels.dp.cpp

```diff
@@ -71,11 +78,1812 @@ namespace dpcpp {
+{
+    queue->submit([&](sycl::handler &cgh) {
+        cgh.parallel_for(sycl_nd_range(grid, block),
+                         [=](sycl::nd_item<3> item_ct1) {
+                             fill_in_dense(num_rows, row_ptrs, col_idxs, values,
+                                           stride, result, item_ct1);
+                         });
+    });
+}
+
+
+template <typename IndexType>
+void calculate_nnz_per_row(size_type num_rows,
+                           const IndexType *__restrict__ row_ptrs,
+                           size_type *__restrict__ nnz_per_row,
+                           sycl::nd_item<3> item_ct1)
+{
+    const auto tidx = thread::get_thread_id_flat(item_ct1);
+    if (tidx < num_rows) {
+        nnz_per_row[tidx] = row_ptrs[tidx + 1] - row_ptrs[tidx];
+    }
+}
+
+template <typename IndexType>
+void calculate_nnz_per_row(dim3 grid, dim3 block,
+                           size_type dynamic_shared_memory, sycl::queue *queue,
+                           size_type num_rows, const IndexType *row_ptrs,
+                           size_type *nnz_per_row)
+{
+    queue->submit([&](sycl::handler &cgh) {
+        cgh.parallel_for(sycl_nd_range(grid, block),
+                         [=](sycl::nd_item<3> item_ct1) {
+                             calculate_nnz_per_row(num_rows, row_ptrs,
+                                                   nnz_per_row, item_ct1);
+                         });
+    });
+}
+
+
+void calculate_slice_lengths(size_type num_rows, size_type slice_size,
+                             size_type stride_factor,
+                             const size_type *__restrict__ nnz_per_row,
+                             size_type *__restrict__ slice_lengths,
+                             size_type *__restrict__ slice_sets,
+                             sycl::nd_item<3> item_ct1)
+{
+    constexpr auto warp_size = config::warp_size;
+    const auto sliceid = item_ct1.get_group(2);
+    const auto tid_in_warp = item_ct1.get_local_id(2);
+
+    if (sliceid * slice_size + tid_in_warp < num_rows) {
+        size_type thread_result = 0;
+        for (int i = tid_in_warp; i < slice_size; i += warp_size) {
+            thread_result =
+                (i + slice_size * sliceid < num_rows)
+                    ? max(thread_result, nnz_per_row[sliceid * slice_size + i])
+                    : thread_result;
+        }
+
+        auto warp_tile = group::tiled_partition<warp_size>(
+            group::this_thread_block(item_ct1));
+        auto warp_result = ::gko::kernels::dpcpp::reduce(
+            warp_tile, thread_result,
+            [](const size_type &a, const size_type &b) { return max(a, b); });
+
+        if (tid_in_warp == 0) {
+            auto slice_length =
+                ceildiv(warp_result, stride_factor) * stride_factor;
+            slice_lengths[sliceid] = slice_length;
+            slice_sets[sliceid] = slice_length;
+        }
+    }
+}
+
+void calculate_slice_lengths(dim3 grid, dim3 block,
```

Comment:
we have a macro for this simple kernel wrapper. something like `hipLaunchKernelGGL` indeed can handle this situation.
but it can not be a general case because some kernels contain shared_memory.
when the kernel contains shared_memory (dynamic or static), there's the place we allocate shared_memory.
adding one more layer for handling it together should work but we already have many macro/layer in dpcpp

Reply:
the template is another issue